mlflow/langchain/api_request_parallel_processor.py

```diff
@@ -218,7 +202,9 @@ def single_call_api(self, callback_handlers: Optional[List[BaseCallbackHandler]]
                 # to maintain existing code, single output chains will still return
                 # only the result
                 response = response.popitem()[1]
-            else:
+            elif not self.stream:
```

Comment:
@BenWilson2 @dbczumar I would like to hear your thoughts on the serialization for the agent streaming response.

The LangChain's pydantic objects, such as `AgentAction` and `AIMessage` are not JSON serializable. The intermediate step contains these objects, so it will cause serialization error in model serving if we do nothing.

For non-stream response, the `_prepare_to_serialize` method addresses this, by simply serialize the response proactively within pyfunc, using LangChain's `dumps` method. At first, I was trying to follow this behavior, but found a few issues in this approach:

**1. Serialized json from `dumps` object has very different structure from the original pydantic object.**

For example, the original `AgentAction` object looks like
```
AgentAction(tool='Search', tool_input='founder of Tesla', log='To answer the question accurately, I will need to perform a search to confirm the current and correct information about the founder of Tesla.\n\nAction: Search\nAction Input: "founder of Tesla"')
```

and `dumps(action)` will be:

```
'{"actions": [{"lc": 1, "type": "constructor", "id": ["langchain", "schema", "agent", "AgentAction"], "kwargs": {"tool": "Search", "tool_input": "founder of Tesla", "log": "To answer the question accurately, I will need to perform a search to confirm the current and correct information about the founder of Tesla.\\n\\nAction: Search\\nAction Input: \\"founder of Tesla\\"", "type": "AgentAction"}}], "messages": [{"lc": 1, "type": "constructor", "id": ["langchain", "schema", "messages", "AIMessage"], "kwargs": {"content": "To answer the question accurately, I will need to perform a search to confirm the current and correct information about the founder of Tesla.\\n\\nAction: Search\\nAction Input: \\"founder of Tesla\\"", "type": "ai", "tool_calls": [], "invalid_tool_calls": []}}]}'
```
This will be confusing, especially for users don't deploy the model on serving, because this implicit serialization has no value. 

**2. For streaming output, the default response is non-json-serializable**
For non-streaming case, the agent only outputs `AgentAction` objects when the agent is created with `return_intermediate_steps=True`. This flag is `False` by default, thus, the response is JSON serializable. The other response fields like `input`, `output` don't have serialization problem. Therefore, the serialization issue only happens for a specific field + a specific opt-in case - so it's ok to do the confusing `dumps` serialization.

On the other hands, streaming response always contains `AgentAction` [by design](https://python.langchain.com/v0.1/docs/modules/agents/how_to/streaming/#stream-intermediate-steps). Users always get the `dumps` serialized response from `predict_stream`. Therefore, the serialization is bigger problem.

**Suggestion**
Therefore, my current thought is to address the serialization problem within model serving, not pyfunc. Similar to how we handled non-json-serializable object for tracing, we can extend the json encoder ([example](https://github.com/databricks/universe/blob/master/model-serving/serving-scheduler/serving-resources/mlflow-serving-server/mlflowserving/scoring_server/scoring_server_utils.py#L64)) to handle these pydantic object.
* The pyfunc `.stream()` response is consistent with LangChain
* The solution works for any LangChain's pydantic model. Most likely `AgentExecutor` is not only one who returns non-json-serializable object, so patching in pyfunc may have risk of overlooking issues in serving.
* We need to update the serving code anyway to handle agent stream output (as discussed offline).

Reply:
1. How many specific pydantic Encoder classes will need to be built within model serving? 
2. How stable are these oddly serialized pydantic APIs? Are there version considerations that will need to be made within serving if we port the encoding serialization functionality within model serving?

Reply:
I agree with the plan here. Let's check with the serving team to see if there is a mechanism that can be indicated whether or not a chain configured for streaming response can issue a helpful message to users in the event that we don't have an appropriate encoder for serialization of a given pydantic class for future feature work on the langchain side.

Reply:
> 1. How many specific pydantic Encoder classes will need to be built within model serving?

We will only have single custom encoder like [this](https://github.com/mlflow/mlflow/blob/4dfb3b839c146a05f0ab4215356d1a33004939fd/mlflow/tracing/utils/__init__.py#L47-L57).

> How stable are these oddly serialized pydantic APIs? Are there version considerations that will need to be made within serving if we port the encoding serialization functionality within model serving?

The serialization API may be updated in major version change, like how Pydantic v1/v2 have different APIs. I don't think the update happens frequently tho - prob. once a year at most.

> I agree with the plan here. Let's check with the serving team to see if there is a mechanism that can be indicated whether or not a chain configured for streaming response can issue a helpful message to users in the event that we don't have an appropriate encoder for serialization of a given pydantic class for future feature work on the langchain side.

Sounds good!